File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.kt
class Solution {
    fun strStr(haystack: String, needle: String): Int {
        val n = haystack.length
        val m = needle.length
        if (m == 0) return 0
        val lps = IntArray(m)
        var len = 0
        var i = 1
        while (i < m) {
            if (needle[i] == needle[len]) {
                len++
                lps[i] = len
                i++
            } else if (len > 0) {
                len = lps[len - 1]
            } else {
                lps[i] = 0
                i++
            }
        }
        i = 0; var j = 0
        while (i < n) {
            if (haystack[i] == needle[j]) {
                i++; j++
                if (j == m) return i - m
            } else if (j > 0) {
                j = lps[j - 1]
            } else {
                i++
            }
        }
        return -1
    }
}

File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.swift
class Solution {
    func strStr(_ haystack: String, _ needle: String) -> Int {
        let n = haystack.count, m = needle.count
        if m == 0 { return 0 }
        let h = Array(haystack), nd = Array(needle)
        var lps = [Int](repeating: 0, count: m)
        var len = 0, i = 1
        while i < m {
            if nd[i] == nd[len] {
                len += 1
                lps[i] = len
                i += 1
            } else if len > 0 {
                len = lps[len - 1]
            } else {
                lps[i] = 0
                i += 1
            }
        }
        i = 0; var j = 0
        while i < n {
            if h[i] == nd[j] {
                i += 1; j += 1
                if j == m { return i - m }
            } else if j > 0 {
                j = lps[j - 1]
            } else {
                i += 1
            }
        }
        return -1
    }
}